### PR TITLE
Base database indices on uid instead of name

### PR DESCRIPTION
**Pull request by robfletcher**

### CODE REVIEW

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ResourceController.kt [hunk 1]
```diff
@@ -62,13 +60,7 @@ class ResourceController(
     // TODO: we need to take the resource type as well so we can actually parse and validate here
     // if you're creating a resource you don't need to pass the name
     log.info("Creating: $submittedResource")
-    val resource = Resource(
-      metadata = ResourceMetadata(name = ResourceName("undefined")),
-      kind = submittedResource.kind,
-      apiVersion = submittedResource.apiVersion,
-      spec = submittedResource.spec
-    )
```
Comment (robfletcher) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ResourceController.kt]: No longer creating a resource with dummy fields here.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/ResourceMetadata.kt [hunk 2]
```diff
@@ -23,17 +23,17 @@ import de.huxhorn.sulky.ulid.ULID
 data class ResourceMetadata(
   val name: ResourceName,
   @JsonProperty(defaultValue = "0") val resourceVersion: Long = 0,
-  val uid: ULID.Value? = null,
+  val uid: ULID.Value,
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/ResourceMetadata.kt]: `uid` no longer nullable.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/ResourceEvent.kt [hunk 3]
```diff
@@ -16,8 +16,13 @@
...
 
 import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.SubmittedResource
+import de.huxhorn.sulky.ulid.ULID
 
-data class ResourceEvent(
-  val type: ResourceEventType,
-  val resource: Resource<*>
-)
+sealed class ResourceEvent
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/ResourceEvent.kt]: Replaced a single event class with a `type` field with 3 different event sub-types as I want each one to have a different payload.

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt [hunk 4]
```diff
@@ -44,20 +58,9 @@ class ClusterHandler(
...
-    return Resource(
-      apiVersion = resource.apiVersion,
-      kind = resource.kind,
-      metadata = metadata,
-      spec = cluster
-    )
-  }
+  override fun generateName(spec: Cluster) = ResourceName(
+    "ec2:cluster:${spec.location.accountName}:${spec.location.region}:${spec.moniker.cluster}"
+  )
```
Comment (robfletcher) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt]: Now each handler just needs to implement a `generateName` and optionally a `validate`. I know we may be moving this logic to a separate type but it still makes sense to simplify it.

File: keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceHandler.kt [hunk 5]
```diff
@@ -29,10 +34,51 @@ interface ResourceHandler<T : Any> : KeelPlugin {
...
+  /**
+   * Validates the resource spec, and generates a metadata header, and applies any defaults /
+   * normalization.
+   *
+   * Implementors should not generally need to override this method.
+   *
+   * @return a hydrated `Resource` with a name generated by convention, a `uid`, default values
+   * applied, etc.
+   */
+  fun validate(resource: SubmittedResource<*>): Resource<T> {
```
Comment (robfletcher) [inline keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceHandler.kt]: The `ResourceHandler` interface can do the default stuff - constructing metadata & parsing the spec. Don't need to re-implement it in every handler.

File: keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceHandler.kt [hunk 6]
```diff
@@ -29,10 +34,51 @@ interface ResourceHandler<T : Any> : KeelPlugin {
...
+   * Validates the resource and applies any defaults / normalization.
+   *
+   * @return [resource] or a copy of [resource] that may have been changed, for example in order to
+   * set default values.
    */
-  fun validate(resource: Resource<*>, generateName: Boolean): Resource<T>
+  @Suppress("UNCHECKED_CAST")
+  fun validate(resource: Resource<Any>): Resource<T> = resource.copy(
+    spec = objectMapper.convertValue(resource.spec, supportedKind.second)
+  ) as Resource<T>
```
Comment (robfletcher) [inline keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceHandler.kt]: This method can be overridden if you want custom validation, normalization, etc.

File: keel-sql/src/main/resources/db/changelog/20190227-initial-schema.yml [hunk 7]
```diff
@@ -53,8 +53,8 @@ databaseChangeLog:
             tableName: resource_state
             columns:
               - column:
-                  name: name
-                  type: varchar(255)
+                  name: uid
+                  type: char(26)
```
Comment (robfletcher) [inline keel-sql/src/main/resources/db/changelog/20190227-initial-schema.yml]: Asher was of the opinion this was a better schema and I think it's more logical to use the primary key of the `resource` table.

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt (hunk 4 above)
Comment (emjburns) [reply keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt]: Makes sense!

**Pull request merged**